api/commands/run.ts

```diff
@@ -413,36 +364,35 @@ export class Runner {
       return;
     }
 
-    if (actionResult.status === dataform.ActionResult.ExecutionStatus.RUNNING) {
-      actionResult.status = dataform.ActionResult.ExecutionStatus.SUCCESSFUL;
-    }
-
     if (
       action.actionDescriptor &&
-      actionResult.status === dataform.ActionResult.ExecutionStatus.SUCCESSFUL &&
+      actionResult.status === dataform.ActionResult.ExecutionStatus.RUNNING &&
       !(this.graph.runConfig && this.graph.runConfig.disableSetMetadata) &&
       action.type === "table" &&
       action.tableType !== "inline"
     ) {
       await this.dbadapter.setMetadata(action);
     }
 
+    let newMetadata: dataform.ITableMetadata;
     if (this.graph.projectConfig.useRunCache) {
-      this.metadataReadPromises.push(
-        (async () => {
-          try {
-            const newMetadata = await this.dbadapter.table(action.target);
-            if (newMetadata) {
-              this.warehouseStateAfterRunByTarget.set(action.target, newMetadata);
-            } else {
-              this.warehouseStateAfterRunByTarget.delete(action.target);
-            }
-          } catch (e) {
-            // If something went wrong trying to get new table metadata, delete it.
-            this.warehouseStateAfterRunByTarget.delete(action.target);
-          }
-        })()
-      );
+      try {
+        newMetadata = await this.dbadapter.table(action.target);
+      } catch (e) {
+        // Ignore Errors thrown when trying to get new table metadata; just allow the relevant
+        // warehouseStateAfterRunByTarget entry to be cleared out (below).
+      }
+    }
+    if (newMetadata) {
+      this.warehouseStateByTarget.set(action.target, newMetadata);
+      actionResult.postExecutionTimestampMillis = newMetadata.lastUpdatedMillis;
+      this.notifyListeners();
+    } else {
+      this.warehouseStateByTarget.delete(action.target);
+    }
+
+    if (actionResult.status === dataform.ActionResult.ExecutionStatus.RUNNING) {
```

Comment:
This seems like a strange bit of code, do we not know this earlier?

Reply:
This is here to prevent us setting metadata if the action has already `FAILED`. I'll add a comment